File: binary_tree.go
package main

import "fmt"

type TreeNode struct {
  value int
  leftChild *TreeNode
  rightChild *TreeNode
}



func insertIntoTree(value int, currentNode *TreeNode) {
  if value < currentNode.value {
    if currentNode.leftChild == nil {
      newNode := TreeNode{}
      newNode.value = value
      currentNode.leftChild = &newNode
    } else {
      insertIntoTree(value, currentNode.leftChild)
    }
  }
  if value > currentNode.value {
    if currentNode.rightChild == nil {
      newNode := TreeNode{}
      newNode.value = value
      currentNode.rightChild = &newNode
    } else {
      insertIntoTree(value, currentNode.rightChild)
    }
  }
}

func main() {

node1 := TreeNode{}
node2 := TreeNode{}
node1.value = 50
node2.value = 75
rootNode := TreeNode{25, &node1, &node2}

fmt.Println(rootNode);

node3 := TreeNode{}
node3.value = 60
insertIntoTree(60, &rootNode)
fmt.Println(rootNode.rightChild.leftChild.value)

}

File: binary_tree.js
class TreeNode {
  constructor(value, left=null, right=null) {
    this.value = value;
    this.leftChild = left;
    this.rightChild = right;
  }
}

function searchTree(searchValue, currentNode) {
  if (currentNode === null || currentNode.value === searchValue) {
    return currentNode
  } else if (searchValue < currentNode.value) {
    return searchTree(searchValue, currentNode.leftChild);
  } else {
    return searchTree(searchValue, currentNode.rightChild);
  }
}

function insertIntoTree(value, currentNode) {
  if (value < currentNode.value) {
    if (!currentNode.leftChild) {
      currentNode.leftChild = TreeNode(value);
    } else {
      insertIntoTree(value, currentNode.leftChild)
    }
  } else if (value > currentNode.value) {
    if (!currentNode.rightChild) {
      currentNode.rightChild = TreeNode(value);
    } else {
      insertIntoTree(value, currentNode.rightChild)
    }
  }
}

function findGreatest(currentNode) {
  if (currentNode.rightChild) {
    return findGreatest(currentNode.rightChild)
  } else {
    return currentNode.rightChild;
  }
}